apps/arweave/c_src/ar_mine_randomx.c

```diff
@@ -910,7 +1026,463 @@ static ERL_NIF_TERM randomx_reencrypt_chunk_nif(
+	uint32_t offset;
+	int jitEnabled, largePagesEnabled, hardwareAESEnabled;
+	struct state* statePtr;
+	ErlNifBinary inputData;
+	ErlNifBinary inputChunk;
 
+	if (argc != 10) {
+		return enif_make_badarg(envPtr);
+	}
+	if (!enif_get_resource(envPtr, argv[0], stateType, (void**) &statePtr)) {
+		return error(envPtr, "failed to read state");
+	}
+	if (!enif_inspect_binary(envPtr, argv[1], &inputData)) {
+		return enif_make_badarg(envPtr);
+	}
+	if (!enif_inspect_binary(envPtr, argv[2], &inputChunk)) {
+		return enif_make_badarg(envPtr);
+	}
+	if (!enif_get_int(envPtr, argv[3], &outChunkLen) ||
+		outChunkLen > MAX_CHUNK_SIZE ||
+		outChunkLen < 64 ||
+		inputChunk.size != outChunkLen ) {
+		return enif_make_badarg(envPtr);
+	}
+	if (!enif_get_int(envPtr, argv[4], &jitEnabled)) {
+		return enif_make_badarg(envPtr);
+	}
+	if (!enif_get_int(envPtr, argv[5], &largePagesEnabled)) {
+		return enif_make_badarg(envPtr);
+	}
+	if (!enif_get_int(envPtr, argv[6], &hardwareAESEnabled)) {
+		return enif_make_badarg(envPtr);
+	}
+	if (!enif_get_int(envPtr, argv[7], &randomxRoundCount)) {
+		return enif_make_badarg(envPtr);
+	}
+	if (!enif_get_int(envPtr, argv[8], &iterations) ||
+		iterations < 1) {
+		return enif_make_badarg(envPtr);
+	}
+	if (!enif_get_uint(envPtr, argv[9], &offset) ||
+		offset < 0 ||
+		offset > MAX_CHUNK_SIZE) {
+		return enif_make_badarg(envPtr);
+	}
+
+	int isRandomxReleased;
+	ERL_NIF_TERM decryptedSubChunkTerm;
+	unsigned char* decryptedSubChunk = enif_make_new_binary(envPtr, outChunkLen,
+			&decryptedSubChunkTerm);
+	uint32_t subChunkSize = outChunkLen;
+	unsigned char key[PACKING_KEY_SIZE];
+
+	randomx_vm *vmPtr = create_vm(statePtr, 1,
+			jitEnabled, largePagesEnabled, hardwareAESEnabled, &isRandomxReleased);
+	if (vmPtr == NULL) {
+		if (isRandomxReleased != 0) {
+			return error(envPtr, "state has been released");
+		}
+		return error(envPtr, "randomx_create_vm failed");
+	}
+
+	unsigned char* subChunk = (unsigned char*)malloc(inputChunk.size);
+	memcpy(subChunk, inputChunk.data, inputChunk.size);
+
+	// 3 bytes is sufficient to represent offsets up to at most MAX_CHUNK_SIZE.
+	int offsetByteSize = 3;
+	unsigned char offsetBytes[offsetByteSize];
+	for (int k = 0; k < offsetByteSize; k++) {
+		offsetBytes[k] = ((offset + subChunkSize) >> (8 * (offsetByteSize - 1 - k))) & 0xFF;
+	}
+	SHA256_CTX sha256;
+	SHA256_Init(&sha256);
+	SHA256_Update(&sha256, inputData.data, inputData.size);
+	SHA256_Update(&sha256, offsetBytes, offsetByteSize);
+	SHA256_Final(key, &sha256);
+
+	for (int j = 0; j < iterations; j++) {
+		randomx_decrypt_chunk(vmPtr, key, PACKING_KEY_SIZE, subChunk, subChunkSize,
+			decryptedSubChunk, randomxRoundCount);
+		if (j < iterations - 1) {
+			memcpy(subChunk, decryptedSubChunk, subChunkSize);
+		}
+	}
+	free(subChunk);
+	destroy_vm(statePtr, vmPtr);
+
+	return ok_tuple(envPtr, decryptedSubChunkTerm);
+}
+
+static ERL_NIF_TERM randomx_reencrypt_legacy_to_composite_chunk_nif(
```

Comment:
The `reencrypt` code that I added back in April 2023 was a mistake, IMO. I'd added it under the assumption that doing the decrypt+encrypt in a single NIF call would improve performance. But once implemented and I could run performance tests the impact was negligible / within the error tolerance of the benchmark: https://digitalhistor-8tx7838.slack.com/archives/C04U7EH3ETF/p1680824283020639

Given that maintaining reencrypt is adding several permutations to the code paths with no real performance gains, my suggestion would be to *not* add them in the new code and instead just use decrypt+encrypt. It will also mean fewer code paths that we have to test, validate, and maintain.

If we still want to go ahead with adding separate reencrypt calls, it might be worth refactoring the code paths to reduce duplication.

Reply:
Hm, yeah, good question. I see you had 1-2% gain there which does not sound like a lot but considering how much packing miners need to do overall and how expensive packing is, I'd say it's worth it. Although the higher the packing difficulty, the lower the gain will be.

Reply:
Got it. Let me revive the benchmark and re-run the tests. If gains still hold I can look into refacoring the duplicate code paths. If my earlier reported gains were wrong, I can pull out the reencrypt implementations.

Reply:
Re-running the benchmarks showed about a 0.5% peformance improvement for reencrypt. We'll keep the reencrypt functions for now, but I'll remove `randomx_reencrypt_legacy_to_composite_chunk_nif` since we'll be using a different randomx library for legacy and composite which makes it hard to put both in the same nif